From 753a83475f09faca05bdba381e8637705fcd30b0 Mon Sep 17 00:00:00 2001
From: Sam Mortimer <sam@mortimer.me.uk>
Date: Wed, 17 Jan 2018 17:25:40 -0800
Subject: [PATCH 1/3] lineage-sdk: Add overlay support for disabling hardware
 features

*) Provide the ability to disable lineage hardware features
   via device tree overlay.

*) Resource integer config_disableHardwareFeatures should
   be the logical OR of the LineageHardwareManager FEATURE_*
   constants for the features that are to be disabled.

*) Useful for disabling features provided by vendor blobs that
   don't actually work properly (eg specific livedisplay libsdm*
   features) and also for testing behavior with features disabled..

Change-Id: I89d556391f37561df691630764b3b2936f4a489e
---
 .../org/lineageos/platform/internal/LineageHardwareService.java   | 3 +++
 lineage/res/res/values/config.xml                                 | 8 ++++++++
 lineage/res/res/values/symbols.xml                                | 3 +++
 3 files changed, 14 insertions(+)

diff --git a/lineage/lib/main/java/org/lineageos/platform/internal/LineageHardwareService.java b/lineage/lib/main/java/org/lineageos/platform/internal/LineageHardwareService.java
index beadbdf..3f1238e 100644
--- a/lineage/lib/main/java/org/lineageos/platform/internal/LineageHardwareService.java
+++ b/lineage/lib/main/java/org/lineageos/platform/internal/LineageHardwareService.java
@@ -168,6 +168,9 @@ public class LineageHardwareService extends LineageSystemService implements Ther
                 mSupportedFeatures |= LineageHardwareManager.FEATURE_PICTURE_ADJUSTMENT;
             if (TouchscreenGestures.isSupported())
                 mSupportedFeatures |= LineageHardwareManager.FEATURE_TOUCHSCREEN_GESTURES;
+            // Remove any features that were requested to be disabled by config.
+            mSupportedFeatures &= ~mContext.getResources().getInteger(
+                    org.lineageos.platform.internal.R.integer.config_disableHardwareFeatures);
         }
 
         public int getSupportedFeatures() {
diff --git a/lineage/res/res/values/config.xml b/lineage/res/res/values/config.xml
index 8374e2b..5c0f60a 100644
--- a/lineage/res/res/values/config.xml
+++ b/lineage/res/res/values/config.xml
@@ -55,6 +55,14 @@
     <!-- BurnIn protection. This should be enabled on devices that have OLED displays -->
     <bool name="config_enableBurnInProtection">false</bool>
 
+    <!-- Lineage hardware features that should be disabled.
+         The value should be the logical OR of LineageHardwareManager
+         FEATURE_* constants.
+         Typical use case is if a vendor blob claims support for
+         something that doesn't actually work.
+         Also useful for testing behavior when features aren't present. -->
+    <integer name="config_disableHardwareFeatures">0</integer>
+
     <!-- Default values for LiveDisplay -->
     <integer name="config_dayColorTemperature">6500</integer>
     <integer name="config_nightColorTemperature">4800</integer>
diff --git a/lineage/res/res/values/symbols.xml b/lineage/res/res/values/symbols.xml
index 6710346..3c9b9a4 100644
--- a/lineage/res/res/values/symbols.xml
+++ b/lineage/res/res/values/symbols.xml
@@ -56,6 +56,9 @@
     <!-- Proximity check on screen on default -->
     <java-symbol type="bool" name="config_proximityCheckOnWakeEnabledByDefault" />
 
+    <!-- Lineage hardware features that should be disabled. -->
+    <java-symbol type="integer" name="config_disableHardwareFeatures" />
+
     <!-- LiveDisplay -->
     <java-symbol type="string" name="live_display_title" />
     <java-symbol type="string" name="live_display_hint" />
-- 
2.7.4

